syft/frameworks/torch/mpc/primitives.py

```diff
@@ -0,0 +1,169 @@
+from syft.exceptions import EmptyCryptoPrimitiveStoreError
+from syft.workers.abstract import AbstractWorker
+
+
+class PrimitiveStorage:
+    """
+    Used by normal workers to store crypto primitives
+    Used by crypto providers to build crypto primitives
+    """
+
+    def __init__(self, owner):
+        # The different kinds of primitives available
+        # Note: each primitive stack is a fixed length list corresponding to all the
+        # components for the primitive. For example, the beaver triple primitive would
+        # have 3 components. Each component is a very multi-dimensional tensor whose
+        # last dimension is the same and corresponds to the number of instances available
+        # for this primitive. That's why get_keys uses a quite complicated dimension
+        # selector. This structure helps generating efficiently primitives using
+        # tensorized key generation algorithms.
+        self.fss_eq: list = []
+        self.fss_comp: list = []
+        self.beaver: list = []
+        self.xor_add_couple: list = []  # couple of the same value shared via ^ or + op
+
+        self._owner: AbstractWorker = owner
+        self._builders: dict = {
+            "fss_eq": self.build_fss_keys(type_op="eq"),
+            "fss_comp": self.build_fss_keys(type_op="comp"),
+            "beaver": self.build_triples,
+            "xor_add_couple": self.build_xor_add_couple,
+        }
+
+    def get_keys(self, type_op, n_instances=1, remove=True):
+        """
+        Return FSS keys primitives #TODO
+
+        Args:
+            type_op: fss_eq, fss_comp, or xor_add_couple
+            n_instances: how many primitives to retrieve. Comparison is pointwise so this in
+                convenient
+            remove: if true, pop out the primitive. If false, only read it. Read mode is
+                needed because we're working on virtual workers and they need to gather
+                a some point and then re-access the keys.
+        """
+        primitive_stack = getattr(self, type_op)
+
+        available_instances = len(primitive_stack[0]) if len(primitive_stack) > 0 else -1
+        if available_instances >= n_instances:
+            keys = []
+            # We iterate on the different elements that constitute a given primitive, for
+            # example of the beaver triples, you would have 3 elements.
+            for i, prim in enumerate(primitive_stack):
+                # We're selecting on the last dimension of the tensor because it's simpler for
+                # generating those primitives in crypto protocols
+                # [:] ~ [slice(None)]
+                # [:1] ~ [slice(1)]
+                # [1:] ~ [slice(1, None)]
+                # [:, :, :1] ~ [slice(None)] * 2 + [slice(1)]
+                n_dim = len(prim.shape)
+                get_slice = [slice(None)] * (n_dim - 1) + [slice(n_instances)]
+                remaining_slice = [slice(None)] * (n_dim - 1) + [slice(n_instances, None)]
```

Comment:
The `torch.narrow()` method didn't do the job?

Reply:
<3 So much better! <3